pkg/security/admission/admission.go

```diff
@@ -0,0 +1,403 @@
+	}
+
+	// remove duplicate constraints and sort
+	matchedConstraints = deduplicateSecurityContextConstraints(matchedConstraints)
+	sort.Sort(ByRestrictions(matchedConstraints))
+	providers, errs := c.createProvidersFromConstraints(a.GetNamespace(), matchedConstraints)
+	logProviders(pod, providers, errs)
+
+	if len(providers) == 0 {
+		return kadmission.NewForbidden(a, fmt.Errorf("no providers available to validated pod request"))
+	}
+
+	// all containers in a single pod must validate under a single provider or we will reject the request
+	validationErrs := fielderrors.ValidationErrorList{}
+	for _, provider := range providers {
+		errs := assignSecurityContext(provider, pod)
+		if len(errs) == 0 {
+			// the entire pod validated, annotate and accept the pod
+			if pod.ObjectMeta.Annotations == nil {
+				pod.ObjectMeta.Annotations = map[string]string{}
+			}
+			glog.V(4).Infof("pod %s (generate: %s) validated against provider %s", pod.Name, pod.GenerateName, provider.GetSCCName())
+			pod.ObjectMeta.Annotations[allocator.ValidatedSCCAnnotation] = provider.GetSCCName()
+			return nil
+		}
+		prefix := fmt.Sprintf("provider %s produced errors:", provider.GetSCCName())
+		validationErrs = append(validationErrs, errs.Prefix(prefix)...)
+	}
+	// we didn't validate against any security context constraint provider, reject the pod and give the
+	// errors for each attempt
+	glog.V(4).Infof("unable to validate pod %s (generate: %s) against any security context constraint: %v", pod.Name, pod.GenerateName, validationErrs)
+	return kadmission.NewForbidden(a, fmt.Errorf("unable to validate against any security context constraint: %v", validationErrs))
+}
+
+// assignSecurityContext creates a security context for each container in the pod
+// and validates that the sc falls within the scc constraints.  All containers must validate against
+// the same scc or is not considered valid.
+func assignSecurityContext(provider scc.SecurityContextConstraintsProvider, pod *kapi.Pod) fielderrors.ValidationErrorList {
+	generatedSCs := make([]*kapi.SecurityContext, len(pod.Spec.Containers))
+
+	for i, c := range pod.Spec.Containers {
+		// Create will create a copy of the SC on the container and use it as the basis for
+		// generation, not touching anything already set.
+		sc, err := provider.CreateSecurityContext(pod, &c)
+		if err != nil {
+			return fielderrors.ValidationErrorList{err}
+		}
+		generatedSCs[i] = sc
+
+		// Validate inspects the container SC and volume information against its SCC.  Here
+		// we will grab the old SC of the container, insert our generated SC and validate.
+		// We will then restore the old SCC.  If we validate the entire pod then we will set
+		// the container SCs back to the generated SC at the end of this method.
+		originalSC := c.SecurityContext
+		c.SecurityContext = sc
+		providerErrs := provider.ValidateSecurityContext(pod, &c)
+		c.SecurityContext = originalSC
```

Comment:
`c` is actually a local struct... you don't have to worry about mutating it's security context. storing the separate `generatedSCs` list is sufficient